Sources/DDGSync/internal/SyncQueue.swift

```diff
@@ -0,0 +1,252 @@
+            partialResult[provider.feature] = provider
+        })
+        self.storage = storage
+        self.crypter = crypter
+        self.getLog = log
+        requestMaker = SyncRequestMaker(storage: storage, api: api, endpoints: endpoints)
+        syncDidFinishPublisher = syncDidFinishSubject.eraseToAnyPublisher()
+        isSyncInProgressPublisher = Publishers
+            .Merge(syncDidStartSubject.map({ true }), syncDidFinishSubject.map({ _ in false }))
+            .prepend(false)
+            .removeDuplicates()
+            .eraseToAnyPublisher()
+    }
+
+    func setUpAndStartFirstSync() async {
+        do {
+            syncDidStartSubject.send(())
+            let syncAuthState = (try? storage.account()?.state) ?? .inactive
+            guard syncAuthState != .inactive else {
+                assertionFailure("Called first sync in unexpected \(syncAuthState) state")
+                return
+            }
+
+            try await withThrowingTaskGroup(of: Void.self) { group in
+                for dataProvider in dataProviders.values {
+                    group.addTask {
+                        try await dataProvider.prepareForFirstSync()
+                    }
+                }
+                for try await _ in group {}
+            }
+
+            if syncAuthState == .addNewDevice {
+                try await sync(fetchOnly: true)
+            }
+
+            if let account = try storage.account()?.updatingState(.active) {
+                try storage.persistAccount(account)
+            }
+
+            try await sync(fetchOnly: false)
+            syncDidFinishSubject.send(.success(()))
+
+        } catch {
+            syncDidFinishSubject.send(.failure(error))
+        }
+    }
+
+    func startSync() async {
+        do {
+            syncDidStartSubject.send(())
+            try await sync(fetchOnly: false)
+            syncDidFinishSubject.send(.success(()))
+        } catch {
+            syncDidFinishSubject.send(.failure(error))
+        }
+    }
+
+    /**
+     * This is private to SyncQueue, but not marked as such to allow unit testing.
+     */
+    func sync(fetchOnly: Bool) async throws {
+        os_log(.debug, log: log, "Sync Operation Started. Fetch-only: %{public}s", String(fetchOnly))
+        defer {
+            os_log(.debug, log: log, "Sync Operation Finished. Fetch-only: %{public}s", String(fetchOnly))
+        }
+
+        try await withThrowingTaskGroup(of: Void.self) { group in
+            for (feature, dataProvider) in dataProviders {
+                group.addTask { [weak self] in
+                    guard let self else {
+                        return
+                    }
+                    os_log(.debug, log: self.log, "Syncing %{public}s", feature.name)
+
+                    do {
+                        var result: SyncResult = try await self.makeResult(for: dataProvider, fetchOnly: fetchOnly)
```

Comment:
Can we split this into Request and Result, then encapsulate first within the latter?